tools/preflight/preflight.go

```diff
@@ -0,0 +1,729 @@
+		Name:         pod.GetName(),
+		Namespace:    o.Namespace,
+		Timeout:      3 * time.Minute,
+		PodCondition: corev1.PodReady,
+		ClientSet:    kubeEnv.GetClientset(),
+	}
+	respChan := make(chan *wait.Response)
+	o.Logger.Infoln("Waiting for source pod to become ready")
+	go waitOptions.WaitOnPod(respChan)
+	wres := <-respChan
+	if wres.Err != nil {
+		o.Logger.Errorln(string(cross), wres.Err.Error())
+		return err
+	} else if !wres.ReachedCondn {
+		return errors.New("Pod " + pod.GetName() + " hasn't reached into ready state.")
+	}
+
+	op := exec.Options{
+		Namespace: o.Namespace,
+		//Command: []string{"sh", "-c", "kubectl",  "exec -it " + pod.GetName(), " -- nslookup kubernetes.default"},
+		Command:       []string{"nslookup", "kubernetes.default"},
+		PodName:       pod.GetName(),
+		ContainerName: dnsContainerName,
+		Executor:      &exec.DefaultRemoteExecutor{},
+		Config:        kubeEnv.GetRestConfig(),
+		ClientSet:     kubeEnv.GetClientset(),
+	}
+	var res *exec.Response
+	isResolved := false
+	resolutionTryCnt := 0
+	for resolutionTryCnt < 3 {
+		if resolutionTryCnt != 0 {
+			o.Logger.Infoln("Retrying to check dns resolution for service kubernetes.default")
+		}
+		responseChan := make(chan *exec.Response)
+		go op.ExecInContainer(responseChan)
+		res = <-responseChan
+		if res != nil && res.Err != nil {
+			o.Logger.Warnln("exec command failed on " + dnsContainerName + " in pod " + pod.GetName() + " :: " + res.Err.Error())
+		} else {
+			isResolved = true
+			break
+		}
+		resolutionTryCnt++
+		time.Sleep(2 * time.Second)
+	}
+
+	if !isResolved {
+		return errors.New("not able to resolve DNS 'kubernetes.default' service inside pods")
+	}
+
+	return nil
+}
+
+// checkVolumeSnapshot checks if volume snapshot and restore is enabled in the cluster
+func (o *Options) checkVolumeSnapshot() error {
+	var (
+		execOp      exec.Options
+		waitOptions *wait.PodWaitOptions
+	)
+
+	pvc := createVolumeSnapshotPVCSpec(o.StorageClass, o.Namespace)
+	pvc, err := kubeEnv.GetClientset().CoreV1().PersistentVolumeClaims(o.Namespace).Create(context.Background(), pvc, metav1.CreateOptions{})
+	if err != nil {
+		o.Logger.Errorln(err.Error())
+		return err
+	}
+	srcPod := createVolumeSnapshotPodSpec(o.Namespace, pvc.GetName())
+	srcPod, err = kubeEnv.GetClientset().CoreV1().Pods(o.Namespace).Create(context.Background(), srcPod, metav1.CreateOptions{})
+	if err != nil {
+		o.Logger.Errorln(err.Error())
+		return err
+	}
+
+	//  Wait for snapshot pod to become ready.
+	waitOptions = &wait.PodWaitOptions{
+		Name:         srcPod.GetName(),
+		Namespace:    o.Namespace,
+		Timeout:      3 * time.Minute,
+		PodCondition: corev1.PodReady,
+		ClientSet:    kubeEnv.GetClientset(),
+	}
+	o.Logger.Infoln("Waiting for snapshot source pod to become ready")
+	err = waitUntilPodCondition(waitOptions)
+	if err != nil {
+		return errors.New("Pod " + srcPod.GetName() + " hasn't reached into ready state.")
+	}
+	o.Logger.Infoln("Created sources pod and pvc")
+
```

Comment:
check if `/demo/data/sample-file.txt` file is created

Reply:
Yes it is created